src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -6905,4 +6905,83 @@ public object VisitParenExpression(ParenExpressionAst parenExpressionAst)
             return parenExpressionAst.Pipeline.Accept(this);
         }
     }
+
+
+    /// <summary>
+    /// Completes with the property names of the InputObject.
+    /// </summary>
+    public class PropertyNameCompleter : IArgumentCompleter
+    {
+        private readonly string _parameterNameOfInput;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PropertyNameCompleter"/> class.
+        /// </summary>
+        public PropertyNameCompleter()
+        {
+            _parameterNameOfInput = "InputObject";
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PropertyNameCompleter"/> class.
+        /// </summary>
+        /// <param name="parameterNameOfInput">The name of the property of the input object for witch to complete with property names.</param>
+        public PropertyNameCompleter(string parameterNameOfInput)
+        {
+            _parameterNameOfInput = parameterNameOfInput;
+        }
+
+        IEnumerable<CompletionResult> IArgumentCompleter.CompleteArgument(
+            string commandName,
+            string parameterName,
+            string wordToComplete,
+            CommandAst commandAst,
+            IDictionary fakeBoundParameters)
+        {
+            var pipelineAst = commandAst.Parent as PipelineAst;
+            if (pipelineAst == null)
+            {
+                return null;
+            }
+
+            int i;
+            for (i = 0; i < pipelineAst.PipelineElements.Count; i++)
```

Comment:
Have you considered using `Ast.FindAll(...).FirstOrDefault()`?

Reply:
Yes. It's a balance. I try hard not to create extra objects, which Linq tends to do. But it sure make the code uglier and harder to read.

And to be honest, I haven't measured the allocations here. 
But it would mean the allocation of a Func delegate, an AstSearcher, a List for the results, and the enumerator. Granted, the last one is an List.Enumerator which is only on the stack.

I fairly often use Linq in higher level code, but when writing infrastructure like this, I try to go with the boring, ugly, but faster option.